README.md
# HTTP plugin for Craft CMS 3.x

Simply return the HTTP status of an url

![Screenshot](resources/img/plugin-logo.png)

## Requirements

This plugin requires Craft CMS 3.0.0-beta.23 or later.

## Installation

To install the plugin, follow these instructions.

1. Open your terminal and go to your Craft project:

        cd /path/to/project

2. Then tell Composer to load the plugin:

        composer require chasegiunta/craft-http

3. In the Control Panel, go to Settings → Plugins and click the “Install” button for http.

## Using HTTP

Ensure a URL is accessible - use `http.ok`.

```
{% if craft.http.ok('http://localhost:8080/devserver/asset/live.js') %}
```

Output the status of a URL, use `http.status` & `true` for second (echo http status) argument.

```
{{ craft.http.status('https://api.somewebsite.com', true) }}
{# echos "HTTP/1.0 301 Moved Permanently" #}

{% if 200 in craft.http.status('https://unpkg.com/vue@2.5.13', true) %}
    <script src="https://unpkg.com/vue@2.5.13">
{% else %}
    <script src="https://cdnjs.cloudflare.com/ajax/libs/vue/2.5.13/vue.js">
{% endif %}
```

HTTP will follow 301/302 redirects until it hits 200.

NOTE: It's not recommended to use this in production as a fallback detect, as a slow loading URL will block your page from loading. This is mainly for *local development* use where a live dev server hosting your assets may or may not be running. Consider using a javascript fallback solution, like [Fallback.js](http://fallback.io/) in production.

Brought to you by [Chase Giunta](chasegiunta.com)


CHANGELOG.md
# HTTP Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) and this project adheres to [Semantic Versioning](http://semver.org/).

## 1.0.0 - 2018-01-14
### Added
- Initial release


src/Http.php
<?php
/**
 * http plugin for Craft CMS 3.x
 *
 * simply return the http status of an asset/url
 *
 * @link      chasegiunta.com
 * @copyright Copyright (c) 2018 Chase Giunta
 */

namespace chasegiunta\http;

use chasegiunta\http\variables\HttpVariable;

use Craft;
use craft\base\Plugin;
use craft\services\Plugins;
use craft\events\PluginEvent;
use craft\web\twig\variables\CraftVariable;

use yii\base\Event;

/**
 * Craft plugins are very much like little applications in and of themselves. We’ve made
 * it as simple as we can, but the training wheels are off. A little prior knowledge is
 * going to be required to write a plugin.
 *
 * For the purposes of the plugin docs, we’re going to assume that you know PHP and SQL,
 * as well as some semi-advanced concepts like object-oriented programming and PHP namespaces.
 *
 * https://craftcms.com/docs/plugins/introduction
 *
 * @author    Chase Giunta
 * @package   Http
 * @since     1.0.0
 *
 */
class Http extends Plugin
{
    // Static Properties
    // =========================================================================

    /**
     * Static property that is an instance of this plugin class so that it can be accessed via
     * Http::$plugin
     *
     * @var Http
     */
    public static $plugin;

    // Public Methods
    // =========================================================================

    /**
     * Set our $plugin static property to this class so that it can be accessed via
     * Http::$plugin
     *
     * Called after the plugin class is instantiated; do any one-time initialization
     * here such as hooks and events.
     *
     * If you have a '/vendor/autoload.php' file, it will be loaded for you automatically;
     * you do not need to load it in your init() method.
     *
     */
    public function init()
    {
        parent::init();
        self::$plugin = $this;

        // Register our variables
        Event::on(
            CraftVariable::class,
            CraftVariable::EVENT_INIT,
            function (Event $event) {
                /** @var CraftVariable $variable */
                $variable = $event->sender;
                $variable->set('http', HttpVariable::class);
            }
        );

        // Do something after we're installed
        Event::on(
            Plugins::class,
            Plugins::EVENT_AFTER_INSTALL_PLUGIN,
            function (PluginEvent $event) {
                if ($event->plugin === $this) {
                    // We were just installed
                }
            }
        );

/**
 * Logging in Craft involves using one of the following methods:
 *
 * Craft::trace(): record a message to trace how a piece of code runs. This is mainly for development use.
 * Craft::info(): record a message that conveys some useful information.
 * Craft::warning(): record a warning message that indicates something unexpected has happened.
 * Craft::error(): record a fatal error that should be investigated as soon as possible.
 *
 * Unless `devMode` is on, only Craft::warning() & Craft::error() will log to `craft/storage/logs/web.log`
 *
 * It's recommended that you pass in the magic constant `__METHOD__` as the second parameter, which sets
 * the category to the method (prefixed with the fully qualified class name) where the constant appears.
 *
 * To enable the Yii debug toolbar, go to your user account in the AdminCP and check the
 * [] Show the debug toolbar on the front end & [] Show the debug toolbar on the Control Panel
 *
 * http://www.yiiframework.com/doc-2.0/guide-runtime-logging.html
 */
        Craft::info(
            Craft::t(
                'http',
                '{name} plugin loaded',
                ['name' => $this->name]
            ),
            __METHOD__
        );
    }

    // Protected Methods
    // =========================================================================

}


src/variables/HttpVariable.php
<?php
/**
 * http plugin for Craft CMS 3.x
 *
 * simply return the http status of an asset/url
 *
 * @link      chasegiunta.com
 * @copyright Copyright (c) 2018 Chase Giunta
 */

namespace chasegiunta\http\variables;

use chasegiunta\http\Http;
use yii\web\BadRequestHttpException;
use yii\base\InvalidConfigException;

use Craft;

/**
 * http Variable
 *
 * Craft allows plugins to provide their own template variables, accessible from
 * the {{ craft }} global variable (e.g. {{ craft.http }}).
 *
 * https://craftcms.com/docs/plugins/variables
 *
 * @author    Chase Giunta
 * @package   Http
 * @since     1.0.0
 */
class HttpVariable
{
    // Public Methods
    // =========================================================================

    /**
     * Check if URL returns '200 OK'
     *
     * @param string $url to check against
     * @param boolean $output whether to output raw http status
     * @return bool|string
     */
    public function status(string $url = null, bool $output = false)
    {
        if ($url === null) {
            throw new InvalidConfigException('URL has not been set on http.status()');
        }
        
        $host = '';
        try {

            if (substr($url, 0, 4) !== 'http' && substr($url, 0, 1 === '/')) {
                // $url seems to be a local URL
                $host = Craft::$app->request->getHostInfo();
            }

            try {
                
                // Fix for SSL (https://stackoverflow.com/a/40830829/4565664)
                stream_context_set_default( [
                    'ssl' => [
                        'verify_peer' => false,
                        'verify_peer_name' => false,
                    ],
                ]);

                $headers = get_headers($host.$url);
            } catch (\Exception $e) {
                // URL offline, not accessible, etc.
                if ($output === true) {
                    throw new BadRequestHttpException('Problem getting headers from '.$host.$url);
                } else {
                    return false;
                }
            }

            if ( strstr($headers[0], '200 OK') ) {

                return $output === true ? $headers[0] : true;

            } else {
                
                if ($output === true) {

                    return $headers[0];
                } else if (preg_match('/^HTTP\/\d\.\d\s+(301|302)/',$headers[0])) {
                    // Test for 301 or 302
                    $url = $this->handleRedirect($url, $headers);

                    return $this->status($url, $output);
                }

            }
            
        } catch (\Exception $e) {
            exit('Something went wrong: ' . $e->getMessage());
        }
    }

    public function ok(string $url = null) {
        return $this->status($url);
    }

    /**
     * Handle 301/302 Redirects
     * 
     * Use cases:
     * http://unpkg.com/vue -> http://unpkg.com/vue (301 redirect)
     * https://unpkg.com/vue -> https://unpkg.com/vue@2.5.13/dist/vue.js (302 redirect)
     *
     * @param string $url
     * @param array $headers
     * @return string URL which should result in a 200 lookup 
     */
    public function handleRedirect(string $url, array $headers)
    {

        foreach($headers as $value)
        {
            if(substr(strtolower($value), 0, 9) === "location:")
            {
                $redirectUrl = trim(substr($value, 9, strlen($value)));
                break;
            }
        }

        // Make sure $redirectUrl is set
        if (!isset($redirectUrl)) {
            return $output === true ? $headers[0].' (failed to follow redirect)' : true;
        }

        // Handle simple HTTP -> HTTPS redirects
        if ( substr($redirectUrl, 0, 5) === 'https' && substr($url, 0, 5) === 'http:' ) {
            $url = str_replace('http:', 'https:', $url);

            return $url;
        }

        // If the $redirectUrl begins a '/', it's relative to the domain of $url
        if (substr($redirectUrl, 0, 1) === "/") {
            $parsedUrl = parse_url($url);
            $redirectUrl = $parsedUrl['scheme'].'://'.$parsedUrl['host'].$redirectUrl;

            return $redirectUrl;
        }
        
    }
}
